node/pkg/driver/node_test.go

```diff
@@ -858,6 +858,90 @@ func TestNodeGetVolumeStats(t *testing.T) {
 	}
 }
 
+func newTestNodeServiceExpand(nodeUtils driver.NodeUtilsInterface, osDevConHelper device_connectivity.OsDeviceConnectivityHelperScsiGenericInterface, nodeMounter driver.NodeMounter) driver.NodeService {
+	return driver.NodeService{
+		Hostname:                    "test-host",
+		ConfigYaml:                  driver.ConfigFile{},
+		VolumeIdLocksMap:            driver.NewSyncLock(),
+		NodeUtils:                   nodeUtils,
+		OsDeviceConnectivityMapping: map[string]device_connectivity.OsDeviceConnectivityInterface{},
+		OsDeviceConnectivityHelper:  osDevConHelper,
+		Mounter:                     nodeMounter,
+	}
+}
+
+func TestNodeExpandVolume(t *testing.T) {
+	d := newTestNodeService(nil, nil)
+	targetPath := "/test/path"
+	volId := "vol-test"
+	expandRequest := &csi.NodeExpandVolumeRequest{
+		VolumeId:   volId,
+		VolumePath: targetPath,
+	}
+	mpathDeviceName := "dm-2"
+	rawSysDevices := "/dev/d1,/dev/d2"
+	sysDevices := []string{"/dev/d1", "/dev/d2"}
+	mpathDevice := "/dev/" + mpathDeviceName
+	fsType := "ext4"
+
+	testCases := []struct {
+		name     string
+		testFunc func(t *testing.T)
+	}{
+		{
+			name: "fail no VolumeId",
+			testFunc: func(t *testing.T) {
+				node := d
+				expandRequest := &csi.NodeExpandVolumeRequest{
+					VolumePath: targetPath,
+				}
+
+				_, err := node.NodeExpandVolume(context.TODO(), expandRequest)
+				assertError(t, err, codes.InvalidArgument)
+			},
+		},
+		{
+			name: "fail no VolumePath",
+			testFunc: func(t *testing.T) {
+				node := d
+				expandRequest := &csi.NodeExpandVolumeRequest{
+					VolumeId: volId,
+				}
+
+				_, err := node.NodeExpandVolume(context.TODO(), expandRequest)
+				assertError(t, err, codes.InvalidArgument)
+			},
+		},
+		{
+			name: "success expand volume",
+			testFunc: func(t *testing.T) {
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+				mockNodeUtils := mocks.NewMockNodeUtilsInterface(mockCtl)
+				mockOsDeviceConHelper := mocks.NewMockOsDeviceConnectivityHelperScsiGenericInterface(mockCtl)
+				mockMounter := mocks.NewMockNodeMounter(mockCtl)
+				node := newTestNodeServiceExpand(mockNodeUtils, mockOsDeviceConHelper, mockMounter)
+
+				mockOsDeviceConHelper.EXPECT().GetMpathDevice(volId).Return(mpathDevice, nil)
+				mockNodeUtils.EXPECT().GetSysDevicesFromMpath(mpathDeviceName).Return(rawSysDevices, nil)
+				mockNodeUtils.EXPECT().RescanPhysicalDevices(sysDevices)
+				mockNodeUtils.EXPECT().ExpandMpathDevice(mpathDeviceName)
+				mockMounter.EXPECT().GetDiskFormat(mpathDevice).Return(fsType, nil)
+				mockNodeUtils.EXPECT().ExpandFilesystem(mpathDevice, fsType)
+
+				_, err := node.NodeExpandVolume(context.TODO(), expandRequest)
+				if err != nil {
+					t.Fatalf("Expect no error but got: %v", err)
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, tc.testFunc)
+	}
```

Comment:
no need for host exception tests? like "no space" or "path not found"?

Reply:
I'm not sure where we could get "no space".
which "path"? I now added a multipath device expansion failure unit test

Reply:
it was an example, I dont know if you can get it, there is several place that `internal grpc code` returned in `NodeExpandVolume`, are they all covered in UT?

Reply:
got it. I now added the rest of them. hope I didn't miss any